### PR TITLE
add take frontend

### PR DESCRIPTION
**Pull request by dayanandasiet**
* add take frontend
* Review commment fix

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by tqchen** [comment]
need to rebase against master, please request reviews from reviewers

**Comment by tqchen** [comment]
@yzhliu @merrymercy can you review this code?

File: nnvm/src/top/tensor/transform.cc [hunk 1]
```diff
@@ -879,5 +879,108 @@ Examples::
...
+
+// take
+DMLC_REGISTER_PARAMETER(TakeParam);
+
+inline bool TakeInferShape(const NodeAttrs& attrs,
+                           std::vector<TShape>* in_shape,
+                           std::vector<TShape>* out_shape) {
+  const TShape& dshape = (*in_shape)[0];
+  const TShape& indicesshape = (*in_shape)[1];
+  if (dshape.ndim() == 0) return false;
```
[the remark this answers is not in the document]
Comment (dayanandasiet) [reply nnvm/src/top/tensor/transform.cc]: return false will throw shape error while compiling operator.

File: nnvm/src/top/tensor/transform.cc [hunk 2]
```diff
@@ -879,5 +879,108 @@ Examples::
...
+}
+
+NNVM_REGISTER_OP(take)
+.describe(R"code(Take elements from an array along an axis.
+
+  When axis is not None, this function does the same thing as 'fancy' indexing
+  (indexing arrays using arrays); however, it can be easier to use if you need
+  elements along a given axis.
+
+  **Note** that when axis is none the flattened input array is used.
```
[the remark this answers is not in the document]
Comment (dayanandasiet) [reply nnvm/src/top/tensor/transform.cc]: yes, added with axis and without axis example on description

File: nnvm/src/top/tensor/transform.cc [hunk 3]
```diff
@@ -879,5 +879,108 @@ Examples::
...
+
+  )code" NNVM_ADD_FILELINE)
+.add_argument("data", "Tensor", "Array to be indexed")
```
[the remark this answers is not in the document]
Comment (dayanandasiet) [reply nnvm/src/top/tensor/transform.cc]: added indices second argument.

**Comment by dayanandasiet** [comment]
@srkreddy1238 

Thanks for your valuable review comments, fixed the comments.

**Comment by dayanandasiet** [comment]
@yzhliu  & @merrymercy 

welcome for review and comments, thanks!

File: nnvm/tests/python/compiler/test_top_level1.py [hunk 4]
```diff
@@ -365,6 +365,46 @@ def test_strided_slice():
...
+        shape_dict = {"a":src_shape, "indices":indices_src.shape}
+        type_dict = {"a":src_dtype, "indices":indices_dtype}
+        graph, lib, _ = nnvm.compiler.build(y, target, shape=shape_dict, dtype=type_dict)
+        m = graph_runtime.create(graph, lib, ctx)
+
+        shape_size = 1
+        for i in range(len(src_shape)):
+            shape_size = shape_size * src_shape[i]
+        a_src = np.arange(shape_size, dtype=src_dtype).reshape((src_shape))
+        if axis is None:
```
Comment (merrymercy) [inline nnvm/tests/python/compiler/test_top_level1.py]: This if is redundant. `np.take(a_src, indices_src, axis=axis)` can cover both case

File: nnvm/tests/python/compiler/test_top_level1.py [hunk 5]
```diff
@@ -365,6 +365,46 @@ def test_strided_slice():
     verify_strided_slice((3, 4, 3), [1, 1, 0], [4, 4])
     verify_strided_slice((3, 4, 3), [1, 1], [4, 4, 3])
 
+def verify_take(src_shape, indices_src, axis=None):
+    src_dtype = "float32"
+    indices_dtype = "int32"
+    indices_src = np.array(indices_src, dtype=indices_dtype)
+    a = sym.Variable("a")
+    indices = sym.Variable("indices")
+    if axis is None:
```
Comment (merrymercy) [inline nnvm/tests/python/compiler/test_top_level1.py]: redundant if statement

**Comment by dayanandasiet** [comment]
@merrymercy 

thanks for your review comments, all review comment is accepted and fixed.

**Comment by tqchen** [comment]
Thanks @merrymercy @srkreddy1238 for the reviews and @dayanandasiet for the contribution, this is now merged!

**Pull request merged**